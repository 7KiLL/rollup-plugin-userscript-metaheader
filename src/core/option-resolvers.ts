export function resolveArrayOption(key: string, values: string[]): string[] {
    if (!values || values.length === 0) {
        return [];
    }
    
    if (!key) {
        throw new Error("Input 'key' cannot be an empty string");
    }
    
    return values.map(value => resolveBaseOption(key, value));
}

export function resolveObjectOption(key: string, values: {[key: string]: string|boolean}) {
    if (!values || Object.keys(values).length === 0) {
        return [];
    }
    
    const options = Object.entries(values).reduce((acc, [objKey, objVal]) => {
        let value;
        if (typeof objVal === 'boolean') {
            value = objKey.trim();
        } else {
            value = `${objKey.trim()}\t${objVal.trim()}`;
        }
        const option = resolveBaseOption(key, value);
        acc.push(option);
        return acc;
    }, [] as string[]);
    
    return options;
}

export function resolveBaseOption(key: string, value: string | boolean | number): string {
    if (!key) {
        throw new Error("Key cannot be null, undefined or an empty string");
    }
    if (value === null || value === undefined) {
        throw new Error("Value cannot be null or undefined");
    }
    if (!['string', 'number'].includes(typeof key)) {
        throw new Error("Invalid type for key. Only string, and number are accepted");
    }
    if (typeof value === 'boolean') {
        return `// @${key.toString().trim()}`
    }
    return `// @${key.toString().trim()}\t${value.toString().trim()}`
}