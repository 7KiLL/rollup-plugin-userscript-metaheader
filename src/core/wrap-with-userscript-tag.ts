export function wrapWithUserscriptTag(str: string) {
    return `// ==UserScript==\n${str}\n// ==/UserScript==\n`
}