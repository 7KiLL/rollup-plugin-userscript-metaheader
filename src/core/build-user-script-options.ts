import {UserScriptHeaderOptions} from "../types/user-script-header-options";
import {resolveArrayOption, resolveBaseOption, resolveObjectOption} from "./option-resolvers";

export function buildUserScriptOptions(headers: UserScriptHeaderOptions): string {
    const meta: string[] = [];
    Object.entries(headers).map(([key, value]) => {
        if (Array.isArray(value)) {
            meta.push(...resolveArrayOption(key, value));
            return;
        }
        if (typeof value === 'object') {
            meta.push(...resolveObjectOption(key, value))
            return;
        }
        meta.push(resolveBaseOption(key, value))
    });
    return meta.join("\n")
}