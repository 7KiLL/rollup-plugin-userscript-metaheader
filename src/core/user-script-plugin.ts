import {Plugin} from "vite";
import {buildUserScriptOptions} from "./build-user-script-options";
import {wrapWithUserscriptTag} from "./wrap-with-userscript-tag";
import {UserScriptPluginOptions} from "../types/user-script-plugin-options";

export function userScriptPlugin(options: UserScriptPluginOptions): Plugin {
    const userScriptOptionsString = buildUserScriptOptions(options.headers);
    const userScriptConfig = wrapWithUserscriptTag(userScriptOptionsString);
    return {
        name: 'add-userscript-metaheader',
        generateBundle(_options, bundle) {
            let file;
            for (const fileName in bundle) {
                const chunk = bundle[fileName];
                if ('isEntry' in chunk && chunk.isEntry) {
                    file = chunk
                    break;
                }
            }

            if (file && 'code' in file) {
                file.code = userScriptConfig + file.code;
            }
        }
    }
}