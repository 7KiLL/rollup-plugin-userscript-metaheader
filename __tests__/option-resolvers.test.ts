/*
Code Analysis

Objective:
The resolveBaseOption function is designed to generate a string option based on a given key and value. It can handle both string and boolean values and formats them accordingly.

Inputs:
- key: a string representing the option key
- value: a string or boolean representing the option value

Flow:
- Check if the value is a boolean
- If it is, return a string with the key and a comment indicating it is a boolean option
- If it is not, return a string with the key, a tab character, and the value, all formatted as a comment

Outputs:
- A string representing the option, formatted as a comment

Additional aspects:
- The function trims the key and value before formatting them
- The function is used by the resolveArrayOption and resolveObjectOption functions to generate options for arrays and objects, respectively.
*/



import {describe, expect, it} from "vitest";
import {resolveArrayOption, resolveBaseOption, resolveObjectOption} from "../src/core/option-resolvers";

describe('resolveBaseOption_function', () => {
    it('test_happy_path_string_value', () => {
        const key = 'key';
        const value = 'value';
        const expected = `// @${key.trim()}	${value.trim()}`;
        const result = resolveBaseOption(key, value);
        expect(result).toBe(expected);
    });

    it('test_happy_path_number_value', () => {
        const key = 'version';
        const value = 1;
        const expected = `// @${key.trim()}	1`;
        const result = resolveBaseOption(key, value);
        expect(result).toBe(expected);
    });
    it('test_happy_path_boolean_value', () => {
        const key = 'key';
        const value = true;
        const expected = `// @${key.trim()}`;
        const result = resolveBaseOption(key, value);
        expect(result).toBe(expected);
    });
    it('test_edge_case_null_key', () => {
        const key = null;
        const value = 'value';
        // @ts-ignore
        expect(() => resolveBaseOption(key, value)).toThrowErrorMatchingSnapshot();
    });
    it('test_edge_case_null_key', () => {
        const key = null;
        const value = 'value';
        // @ts-ignore
        expect(() => resolveBaseOption(key, value)).toThrowErrorMatchingSnapshot()
    });

    it('test_edge_case_numeric_key', function () {
        const key = 1;
        const value = 'any'
        // @ts-ignore
        expect(resolveBaseOption(key, value)).toMatchSnapshot();
    });

    it('test_edge_case_object_key', function () {
        const key = {};
        const value = 'value';
        // @ts-ignore
        expect(() => resolveBaseOption(key, value)).toThrowErrorMatchingSnapshot()
    });

    it('test_edge_case_array_key', function () {
        const key: unknown[] = [];
        const value = 'value';
        // @ts-ignore
        expect(() => resolveBaseOption(key, value)).toThrowErrorMatchingSnapshot()
    });
});

describe('resolveArrayOption_function', () => {
    it('test_empty_values', () => {
        const key = 'key'
        const values: string[] = []
        const expected: string[] = []
        const result = resolveArrayOption(key, values)
        expect(result).toEqual(expected)
    })

    it('test_string_values', () => {
        const key = 'key'
        const values = ['value1', 'value2']
        const expected = ['// @key\tvalue1', '// @key\tvalue2']
        const result = resolveArrayOption(key, values)
        expect(result).toEqual(expected)
    })
    it('test_boolean_values', () => {
        const key = 'key'
        const values = [true, false]
        const expected = ['// @key', '// @key']
        // @ts-ignore
        const result = resolveArrayOption(key, values)
        expect(result).toEqual(expected)
    })

    it('test_null_key', () => {
        const key = null
        const values = ['value1', 'value2']
        // @ts-ignore
        expect(() => resolveArrayOption(key, values)).toThrowErrorMatchingSnapshot()
    })

    it('test_null_values', () => {
        const key = 'key'
        const values = null
        // @ts-ignore
        expect(resolveArrayOption(key, values)).toEqual([])
    })

    it('test_invalid_values', () => {
        const key = 'key'
        const values = [1, 2]
        // @ts-ignore
        expect(resolveArrayOption(key, values)).toEqual(["// @key\t1","// @key\t2"])
    })
})

describe('resolveObjectOption_function', () => {
    it('test_empty_values_object', () => {
        const key = 'test_key';
        const values = {};
        const result = resolveObjectOption(key, values);
        expect(result).toEqual([]);
    });

    it('test_valid_key_and_values_object', () => {
        const key = 'test_key';
        const values = {
            'value1': 'option1',
            'value2': 'option2'
        };
        const result = resolveObjectOption(key, values);
        expect(result).toEqual([
            '// @test_key\tvalue1\toption1',
            '// @test_key\tvalue2\toption2'
        ]);
    });
    it('test_null_key', () => {
        const key = null;
        const values = {
            'value1': 'option1',
            'value2': 'option2'
        };
        // @ts-ignore
        expect(() => resolveObjectOption(key, values)).toThrowErrorMatchingSnapshot();
    });
    it('test_undefined_key', () => {
        const key = undefined;
        const values = {
            'value1': 'option1',
            'value2': 'option2'
        };
        // @ts-ignore
        expect(() => resolveObjectOption(key, values)).toThrowErrorMatchingSnapshot();
    });
    it('test_boolean_object', () => {
        const key = 'key';
        const values = {
            'value1': true,
            'value2': true
        };
        // @ts-ignore
        expect(resolveObjectOption(key, values)).toMatchSnapshot();
    });
    it('test_null_values_object', () => {
        const key = 'test_key';
        const values = null;
        // @ts-ignore
        expect(resolveObjectOption(key, values)).toEqual([]);
    });
    it('test_undefined_values_object', () => {
        const key = 'test_key';
        const values = undefined;
        // @ts-ignore
        expect(resolveObjectOption(key, values)).toEqual([]);
    });
})

