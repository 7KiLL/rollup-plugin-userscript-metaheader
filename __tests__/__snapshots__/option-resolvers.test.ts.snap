// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolveArrayOption_function > test_null_key 1`] = `"Input 'key' cannot be an empty string"`;

exports[`resolveBaseOption_function > test_edge_case_array_key 1`] = `"Invalid type for key. Only string, and number are accepted"`;

exports[`resolveBaseOption_function > test_edge_case_null_key 1`] = `"Key cannot be null, undefined or an empty string"`;

exports[`resolveBaseOption_function > test_edge_case_null_key 2`] = `"Key cannot be null, undefined or an empty string"`;

exports[`resolveBaseOption_function > test_edge_case_numeric_key 1`] = `"// @1	any"`;

exports[`resolveBaseOption_function > test_edge_case_object_key 1`] = `"Invalid type for key. Only string, and number are accepted"`;

exports[`resolveObjectOption_function > test_boolean_object 1`] = `
[
  "// @key	value1",
  "// @key	value2",
]
`;

exports[`resolveObjectOption_function > test_null_key 1`] = `"Key cannot be null, undefined or an empty string"`;

exports[`resolveObjectOption_function > test_undefined_key 1`] = `"Key cannot be null, undefined or an empty string"`;
